core/src/msg.rs

```diff
@@ -0,0 +1,352 @@
+//! Message queue management.
+
+use alloc::{collections::BTreeMap, sync::Arc, vec::Vec};
+
+use axerrno::{AxError, AxResult};
+use axhal::time::monotonic_time_nanos;
+use axsync::Mutex;
+use linux_raw_sys::general::*;
+use starry_process::Pid;
+
+use super::shm::IpcPerm;
+
+/// Data structure describing a message queue.
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub struct MsqidDs {
+    /// operation permission struct
+    pub msg_perm: IpcPerm,
+    /// time of last msgsnd()
+    pub msg_stime: __kernel_time_t,
+    /// time of last msgrcv()
+    pub msg_rtime: __kernel_time_t,
+    /// time of last change by msgctl()
+    pub msg_ctime: __kernel_time_t,
+    /// current number of bytes on queue
+    pub msg_cbytes: __kernel_size_t,
+    /// number of messages in queue
+    pub msg_qnum: __kernel_size_t,
+    /// max number of bytes on queue
+    pub msg_qbytes: __kernel_size_t,
+    /// pid of last msgsnd()
+    pub msg_lspid: __kernel_pid_t,
+    /// pid of last msgrcv()
+    pub msg_lrpid: __kernel_pid_t,
+}
+
+impl MsqidDs {
+    fn new(key: i32, mode: __kernel_mode_t, pid: __kernel_pid_t, uid: u32, gid: u32) -> Self {
+        Self {
+            msg_perm: IpcPerm {
+                key,
+                uid,
+                gid,
+                cuid: uid,
+                cgid: gid,
+                mode,
+                seq: 0,
+                pad: 0,
+                unused0: 0,
+                unused1: 0,
+            },
+            msg_stime: 0,
+            msg_rtime: 0,
+            msg_ctime: monotonic_time_nanos() as __kernel_time_t,
+            msg_cbytes: 0,
+            msg_qnum: 0,
+            msg_qbytes: MSGMNB as __kernel_size_t,
+            msg_lspid: pid,
+            msg_lrpid: pid,
+        }
+    }
+}
+
+/// Single message in the queue
+pub struct Message {
+    /// message type
+    pub mtype: i64,
+    /// message data
+    pub data: Vec<u8>,
+    /// timestamp when message was sent
+    pub timestamp: u64,
+}
+
+/// This struct is used to maintain the message queue in kernel.
+pub struct MsgInner {
```

Comment:
Change the name to reflect the `queue`.

Reply:
And many other places using the name of `inner`, including `msg_inner`, `get_inner_by_msqid`, `insert_msqid_inner`.